### PR TITLE
[SPARK-27338][Core] Fix deadlock in UnsafeExternalSorter.SpillableIterator when locking both UnsafeExternalSorter.SpillableIterator and TaskMemoryManager

### PR DESCRIPTION
**Pull request by venkata91**
## What changes were proposed in this pull request?

In `UnsafeExternalSorter.SpillableIterator#loadNext()` takes lock on the `UnsafeExternalSorter` and calls `freePage` once the `lastPage` is consumed which needs to take a lock on `TaskMemoryManager`. At the same time, there can be another MemoryConsumer using `UnsafeExternalSorter` as part of sorting can try to `allocatePage` needs to get lock on `TaskMemoryManager` which can cause spill to happen which requires lock on `UnsafeExternalSorter` again causing deadlock. This is a classic deadlock situation happening similar to the SPARK-26265.

To fix this, we can move the `freePage` call in `loadNext` outside of `Synchronized` block similar to the fix in SPARK-26265

## How was this patch tested?

Manual tests were being done and will also try to add a test.

### CODE REVIEW

**Comment by beliefer** [comment]
I checked the code and found `UnsafeExternalSorter.split` will call the `SpillableIterator.split`.
`SpillableIterator.split` need to get lock on `UnsafeExternalSorter`.
`UnsafeExternalSorter.loadNext` will call the `MemoryConsumer.freePage` and `MemoryConsumer.freePage` will call the `TaskMemoryManager.freePage`.
`TaskMemoryManager.freePage` need to get lock on `TaskMemoryManager`.
I think your PR seems good!

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java [hunk 1]
```diff
@@ -573,19 +573,26 @@ public boolean hasNext() {
...
     @Override
     public void loadNext() throws IOException {
-      synchronized (this) {
-        loaded = true;
-        if (nextUpstream != null) {
-          // Just consumed the last record from in memory iterator
-          if (lastPage != null) {
-            freePage(lastPage);
-            lastPage = null;
+      MemoryBlock pageToFree = null;
```
Comment (beliefer) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java]: I think your PR seems good!

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java [hunk 2]
```diff
@@ -573,19 +573,26 @@ public boolean hasNext() {
...
+      try {
+        synchronized (this) {
+          loaded = true;
+          if (nextUpstream != null) {
+            // Just consumed the last record from in memory iterator
+            if(lastPage != null) {
+              pageToFree = lastPage;
```
Comment (cloud-fan) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java]: can we add some comment to explain it? You can just copy it from https://github.com/apache/spark/pull/24269/files#diff-027299fb14327ddcaba457f81ecff32cR583

**Review by attilapiros — COMMENTED** [review]
+1 for adding some explanation (the comment mentioned by @cloud-fan) otherwise LGTM

**Comment by cloud-fan** [comment]
ok to test

**Review by jiangxb1987 — APPROVED** [review]
LGTM, please update the comment.

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java (hunk 2 above)
Comment (venkata91) [reply core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java]: Sure @cloud-fan, will add and update the PR

Comment (venkata91) [reply core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java]: Done @cloud-fan . I didn't know that you also filed a PR simultaneously for the same issue. :)

**Comment by cloud-fan** [comment]
thanks, merging to master/2.4/2.3!

**Pull request closed without merging**